File: src/studio/lib/models/criteria.dart
library;

import 'package:flutter/foundation.dart';

/// 考评准则数据模型（v0.1 只读 API）。
///
/// 契约先行：字段以 `data/profile/qtrecurit/criteria.json` 为基准，
/// 与 Provider（`internal/model/criterion.go`）同源。
/// policy 与 assessment 承载长文（content），criterion 承载细则（description）。

/// Policy 是政策：知识提炼链产物的公开化展示。
@immutable
class Policy {
  final String id;
  final String title;
  final String content;
  final String status;

  const Policy({
    required this.id,
    required this.title,
    required this.content,
    required this.status,
  });

  factory Policy.fromJson(Map<String, dynamic> json) => Policy(
    id: json['id'] as String,
    title: json['title'] as String,
    content: json['content'] as String,
    status: (json['status'] as String?) ?? 'active',
  );
}

/// Criterion 是筛选标准：五个检验维度（期望匹配、动机视角、了解程度、学习意向、责任心）的落地细则。
@immutable
class Criterion {
  final String id;
  final String title;
  final String description;
  final String status;

  const Criterion({
    required this.id,
    required this.title,
    required this.description,
    required this.status,
  });

  factory Criterion.fromJson(Map<String, dynamic> json) => Criterion(
    id: json['id'] as String,
    title: json['title'] as String,
    description: json['description'] as String,
    status: (json['status'] as String?) ?? 'active',
  );
}

/// Assessment 是考核说明：考核分层、序列轨道、问卷与 AI 分析说明。
@immutable
class Assessment {
  final String id;
  final String title;
  final String content;
  final String status;

  const Assessment({
    required this.id,
    required this.title,
    required this.content,
    required this.status,
  });

  factory Assessment.fromJson(Map<String, dynamic> json) => Assessment(
    id: json['id'] as String,
    title: json['title'] as String,
    content: json['content'] as String,
    status: (json['status'] as String?) ?? 'active',
  );
}

File: src/provider/internal/api/criteria.go
package api

import (
	"encoding/json"
	"log/slog"
	"net/http"

	"github.com/quanttide/qtrecurit-provider/internal/model"
	"github.com/quanttide/qtrecurit-provider/internal/store"
)

// 考评准则只读 API：政策、筛选标准、考核说明——无需登录即可读，
// 是「考评方式透明公开」的基础。三个资源平级，各自独立。

// PolicyHandler 提供政策只读 API。
type PolicyHandler struct {
	store store.Store
}

func NewPolicyHandler(st store.Store) *PolicyHandler {
	return &PolicyHandler{store: st}
}

func (h *PolicyHandler) ListPolicies(w http.ResponseWriter, r *http.Request) {
	var policies []model.Policy
	list(w, h.store, "policies", &policies)
}

func (h *PolicyHandler) GetPolicy(w http.ResponseWriter, r *http.Request) {
	var policy model.Policy
	get(w, r, h.store, "policies", &policy)
}

// CriterionHandler 提供筛选标准只读 API。
type CriterionHandler struct {
	store store.Store
}

func NewCriterionHandler(st store.Store) *CriterionHandler {
	return &CriterionHandler{store: st}
}

func (h *CriterionHandler) ListCriteria(w http.ResponseWriter, r *http.Request) {
	var criteria []model.Criterion
	list(w, h.store, "criteria", &criteria)
}

func (h *CriterionHandler) GetCriterion(w http.ResponseWriter, r *http.Request) {
	var criterion model.Criterion
	get(w, r, h.store, "criteria", &criterion)
}

// AssessmentHandler 提供考核说明只读 API。
type AssessmentHandler struct {
	store store.Store
}

func NewAssessmentHandler(st store.Store) *AssessmentHandler {
	return &AssessmentHandler{store: st}
}

func (h *AssessmentHandler) ListAssessments(w http.ResponseWriter, r *http.Request) {
	var assessments []model.Assessment
	list(w, h.store, "assessments", &assessments)
}

func (h *AssessmentHandler) GetAssessment(w http.ResponseWriter, r *http.Request) {
	var assessment model.Assessment
	get(w, r, h.store, "assessments", &assessment)
}

func list(w http.ResponseWriter, st store.Store, collection string, target any) {
	data, err := st.List(collection)
	if err != nil {
		slog.Error("list "+collection, "error", err)
		WriteError(w, "INTERNAL_ERROR", "failed to list "+collection, http.StatusInternalServerError)
		return
	}
	if err := json.Unmarshal(data, target); err != nil {
		slog.Error("parse "+collection, "error", err)
		WriteError(w, "INTERNAL_ERROR", "failed to parse "+collection, http.StatusInternalServerError)
		return
	}
	WriteJSON(w, target, http.StatusOK)
}

func get(w http.ResponseWriter, r *http.Request, st store.Store, collection string, target any) {
	id := r.PathValue("id")
	data, err := st.Get(collection, id)
	if err != nil {
		WriteError(w, "NOT_FOUND", "record not found", http.StatusNotFound)
		return
	}
	if err := json.Unmarshal(data, target); err != nil {
		slog.Error("parse "+collection, "error", err)
		WriteError(w, "INTERNAL_ERROR", "failed to parse "+collection, http.StatusInternalServerError)
		return
	}
	WriteJSON(w, target, http.StatusOK)
}
